server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/TransportNodesInfoAction.java

```diff
@@ -97,20 +98,39 @@ protected NodeInfo nodeOperation(NodeInfoRequest nodeRequest, Task task) {
     public static class NodeInfoRequest extends TransportRequest {
 
         NodesInfoRequest request;
+        private NodesInfoMetrics nodesInfoMetrics;
 
         public NodeInfoRequest(StreamInput in) throws IOException {
             super(in);
-            request = new NodesInfoRequest(in);
+            if (in.getTransportVersion().onOrAfter(NODE_INFO_REQUEST_VERSIONS_ADDED)) {
+                this.nodesInfoMetrics = new NodesInfoMetrics(in);
+            } else {
+                this.request = new NodesInfoRequest(in);
+            }
         }
 
         public NodeInfoRequest(NodesInfoRequest request) {
+            this.nodesInfoMetrics = request.getNodesInfoMetrics();
             this.request = request;
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
             super.writeTo(out);
-            request.writeTo(out);
+            if (out.getTransportVersion().onOrAfter(NODE_INFO_REQUEST_VERSIONS_ADDED)) {
+                this.nodesInfoMetrics.writeTo(out);
+            } else {
+                this.request.writeTo(out);
```

Comment:
I think it would work to create a new `NodesInfoRequest` right here just for sending out to older nodes (with no node IDs, just the metrics). That way we wouldn't have to keep track of the request and the metrics separately, because that's going to be problematic if we ever tried to send one of these things over the wire twice with two different transport versions.

Reply:
Big thanks, such a great idea! This way we don't have to worry about backwards compatibility. I have force pushed commits to make them easier to review. [fa7c890](https://github.com/elastic/elasticsearch/pull/99938/commits/fa7c890bd26d74edc73a8813c0c9edda7cfbc21f)